### ISSUE TITLE
Optimal way of accumulating gradients?

### ISSUE THREAD

**Issue by vgoklani** [ml-explore/mlx-examples#929]
Is there an official example of the optimal pattern for accumulating gradients?

Thanks!

**Comment by awni** [comment]
Here's a really toy example of how to do gradient accumulation in MLX. This should work fine for non-toy models as well, but kept it simple just to demonstrate the pieces:

```python
import mlx.core as mx
import mlx.nn as nn
import mlx.optimizers as optim
from mlx.utils import tree_map

model = nn.Linear(10, 10)

batch = mx.random.uniform(shape=(100, 10))

def loss_fn(x):
    return model(x).sum()

loss_and_grad_fn = nn.value_and_grad(model, loss_fn)

optimizer = optim.SGD(learning_rate=1e-5)

num_iterations = 1
for _ in range(num_iterations):
    # One step with gradient accumulation in micro batches of size 8
    accum_grads = None
    step = 8
    for i in range(0, batch.shape[1], step):
        loss, grads = loss_and_grad_fn(batch[i:i+step])
        if accum_grads is not None:
            accum_grads = tree_map(mx.add, grads, accum_grads)
        else:
            accum_grads = grads
        del grads
        mx.eval(accum_grads)

    # Probably good to rescale the grads by the overall batch size
    accum_grads = tree_map(lambda g: (step / batch.shape[1]) * g, accum_grads)

    # Take the optimizer step
    optimizer.update(model, accum_grads)
    mx.eval(optimizer.state)
```
